go/core/action.go

```diff
@@ -28,6 +28,26 @@ import (
 	"github.com/invopop/jsonschema"
 )
 
+// Middleware is a functions that take in the request object, response object,
```

Comment:
In general I think @jba 's goal was for the core package to be for plugins and the genkit and ai packages to be for users.  This makes me wonder who is going to write `Middleware` values: the user or plugins?

Reply:
Middleware is a feature of the core framework at the action level and ai package will contain a couple of common model middlewares (see: https://github.com/firebase/genkit/blob/main/js/ai/src/model/middleware.ts).
There are ideas to potentially implement some retriever middleware as well (for advances RAG patterns).

Reply:
What I mean is: which code is going to pass the slice of middleware functions?  Will an application that uses Genkit be writing its own middleware functions?

Reply:
No, middleware functions are not meant to be written by the application/plugin developer (it's OK if they want to -- middleware is a powerful design pattern), but meant to be used by them. Here's an example of reusable middleware used to add common features to models: https://github.com/firebase/genkit/blob/main/js/plugins/vertexai/src/gemini.ts#L450-L464